pkg/k8s/rbac.go

```diff
@@ -4,43 +4,136 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/keikoproj/iam-manager/constants"
 	"github.com/keikoproj/iam-manager/pkg/log"
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
-
 	corev1 "k8s.io/api/core/v1"
-	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 )
 
-//CreateServiceAccount adds the service account
-func (c *Client) CreateOrUpdateServiceAccount(ctx context.Context, saName string, ns string, roleARN string) error {
-	log := log.Logger(ctx, "pkg.k8s", "rbac", "CreateOrUpdateServiceAccount")
+// ServiceAccountRequest struct
+type ServiceAccountRequest struct {
+	Namespace          string
+	IamRoleARN         string
+	ServiceAccountName string
+}
+
+// EnsureServiceAccount decomposes a ServiceAccountRequest into the right behavior and
+// desired target state of a ServiceAccount<->Iamrole mapping, and implements that.
+//
+// Current functionality is limited - but I am to increase the complexity and control
+// of how this mapping works, which and I want to keep as little complexity in the
+// CreateOrUpdateServiceAccount function as possible.
+func (c *Client) EnsureServiceAccount(ctx context.Context, req ServiceAccountRequest) (*corev1.ServiceAccount, error) {
+	log := log.Logger(ctx, "pkg.k8s", "rbac", "EnsureServiceAccount")
+	log = log.WithValues("request", req)
+
+	// TODO: Implement a check in the future to figure out whether or not we _were supposed_ to manage the
+	// ServiceAccount at one point, and should now _remove_ the annotation from it or not.
+
+	// First thing, check if the SA exists already. If it does, we will avoid a failed Create call.
+	sa, err := c.GetOrCreateServiceAccount(ctx, req.ServiceAccountName, req.Namespace)
+	if err != nil {
+		log.Error(err, "Unable to get or create ServiceAccount!")
+		return nil, err
+	}
+	log.V(1).Info("Got existing SA", "serviceaccount", sa)
+
+	// If the SA exists, let's check its annotations now. If they match, we don't need to do any
+	// more work and we can return!
+	currentVal, _ := sa.Annotations[constants.ServiceAccountRoleAnnotation]
+
+	// Does it match? If so, we're done here. Return.
+	if currentVal == req.IamRoleARN {
```

Comment:
How are you guys using IRSA? Can you have multiple iam roles per namespace? 
Scenario:
1. IamRole1 created with IRSA annotation sa-namespace
2. Pods with sa-namespace now have IamRole1 access
3. IamRole2 create with IRSA annotation with same service account name sa-namespace
4. This code will overwrite with IamRole2 and pods in step2 will lose the access to IamRole1

Our use case is one iam role per namespace so we probably wound't run into this issue but we have to think more here than just updating with new iam role ARN. (Thinking out loud) We either have to decline iamrole creation request stating that Service account is already associated with other iam role. Let me know if you have any ideas?